src/mono/mono/metadata/metadata.c

```diff
@@ -1057,7 +1057,15 @@ mono_metadata_string_heap (MonoImage *meta, guint32 index)
 const char *
 mono_metadata_string_heap_checked (MonoImage *meta, guint32 index, MonoError *error)
 {
-	if (G_UNLIKELY (!(index < meta->heap_strings.size))) {
+	if (mono_image_is_dynamic (meta))
+	{
+		MonoDynamicImage* img = (MonoDynamicImage*) meta;
+		const char *image_name = meta && meta->name ? meta->name : "unknown image";
+		if (G_UNLIKELY (!(index < img->sheap.index)))
+			mono_error_set_bad_image_by_name (error, image_name, "string heap index %ud out bounds %u: %s", index, img->sheap.index, image_name);
+		return img->sheap.data + index;
+	}
```

Comment:
I'm pretty sure you need a similar change to `mono_metadata_blob_heap_checked` - not sure if the test would cover it

Reply:
Done and updated the test